api/datastore/sql/migrations/migrations.go

```diff
@@ -1,534 +1,169 @@
-		rv = append(rv, childName)
-	}
-	return rv, nil
-}
 
-type bintree struct {
-	Func     func() (*asset, error)
-	Children map[string]*bintree
 }
 
-var _bintree = &bintree{nil, map[string]*bintree{
-	"1_add_route_created_at.down.sql": &bintree{_1_add_route_created_atDownSql, map[string]*bintree{}},
-	"1_add_route_created_at.up.sql":   &bintree{_1_add_route_created_atUpSql, map[string]*bintree{}},
-	"2_add_call_stats.down.sql":       &bintree{_2_add_call_statsDownSql, map[string]*bintree{}},
-	"2_add_call_stats.up.sql":         &bintree{_2_add_call_statsUpSql, map[string]*bintree{}},
-	"3_add_call_error.down.sql":       &bintree{_3_add_call_errorDownSql, map[string]*bintree{}},
-	"3_add_call_error.up.sql":         &bintree{_3_add_call_errorUpSql, map[string]*bintree{}},
-	"4_add_route_updated_at.down.sql": &bintree{_4_add_route_updated_atDownSql, map[string]*bintree{}},
-	"4_add_route_updated_at.up.sql":   &bintree{_4_add_route_updated_atUpSql, map[string]*bintree{}},
-	"5_add_app_created_at.down.sql":   &bintree{_5_add_app_created_atDownSql, map[string]*bintree{}},
-	"5_add_app_created_at.up.sql":     &bintree{_5_add_app_created_atUpSql, map[string]*bintree{}},
-	"6_add_app_updated_at.down.sql":   &bintree{_6_add_app_updated_atDownSql, map[string]*bintree{}},
-	"6_add_app_updated_at.up.sql":     &bintree{_6_add_app_updated_atUpSql, map[string]*bintree{}},
-	"7_add_route_cpus.down.sql":       &bintree{_7_add_route_cpusDownSql, map[string]*bintree{}},
-	"7_add_route_cpus.up.sql":         &bintree{_7_add_route_cpusUpSql, map[string]*bintree{}},
-}}
+func ApplyMigrations(ctx context.Context, driver string, db *sql.DB) error {
+	log := common.Logger(ctx)
 
-// RestoreAsset restores an asset under the given directory
-func RestoreAsset(dir, name string) error {
-	data, err := Asset(name)
-	if err != nil {
-		return err
-	}
-	info, err := AssetInfo(name)
-	if err != nil {
-		return err
-	}
-	err = os.MkdirAll(_filePath(dir, filepath.Dir(name)), os.FileMode(0755))
-	if err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(_filePath(dir, name), data, info.Mode())
-	if err != nil {
-		return err
-	}
-	err = os.Chtimes(_filePath(dir, name), info.ModTime(), info.ModTime())
-	if err != nil {
-		return err
+	for _, v := range initialTables {
+		_, err := db.ExecContext(ctx, v)
+		if err != nil {
+			return err
+		}
 	}
-	return nil
-}
 
-// RestoreAssets restores an asset under the given directory recursively
-func RestoreAssets(dir, name string) error {
-	children, err := AssetDir(name)
-	// File
+	goose.SetDialect(driver)
+	migrations = sortAndConnectMigrations(migrations)
+	current, err := goose.GetDBVersion(db)
+	log.Info("current datastore version: ", current)
 	if err != nil {
-		return RestoreAsset(dir, name)
+		if err != goose.ErrNoNextVersion {
+			return err
+		}
 	}
-	// Dir
-	for _, child := range children {
-		err = RestoreAssets(dir, filepath.Join(name, child))
-		if err != nil {
+	// datastore is fresh new
```

Comment:
we also need to check if they are coming from the previous migration package? If we special case errors coming out of there then we can end up with a db that's in a funky state (say some DBA went around changing the tables..)

Reply:
I'm not sure we actually have to worry about that because the migrations we have at this moment check SQL error to make sure that it's a duplicated column error (specific to each datastore we support). So we don't have to worry about previous version table, according to the process of migrations applying (in this patch) there's no point to check that version because we are safe to apply the migrations.

Reply:
regarding DBA, we aren't exposing datastore API to change the datastore schemas, if someone did that - it's his fault.

Reply:
we're running migrations twice, effectively? it kind of smells, I could see that causing issues. we're effectively migrating a database that is up to date - there seems to be a reason to skip these, what with the versioning of each migration and deliberately skipping over migrations that have already run.

Reply:
We're not running migrations twice. If datastore has the latest version we skip the migrations.

Reply:
see `migrations.go`, lines 158 to 166

Reply:
this is only if they were coming from this migrations library? there was a note somewhere that the table that keeps the migration version is different for goose than for the library we were using.

Reply:
that's true, indeed, tables are different, but as noted as well, reading another migrations table is redundant, it brings yet another overhead to the migrations. Basically, nothing would be changed inside Fn tables, but a new table for the migrations would be created.

So the migration from `migrate` framework to `goose` ends up in a new table for the migration versions and zero diffs for the Fn tables.

Reply:
right. we could just read the version from old migrate table, drop that table, and then proceed like normal? this would fix the error handling stuff, and at least doesn't seem too challenging to do but maybe it is?

Reply:
Done, new commit adds corresponding check, see 68a523d.

Reply:
some confirmation that this was tested? it's not worth having automated tests for, imo

Reply:
It was tested, i used the following steps:
1. checkout `master` and build the binary
2. start Fn server with corresponding datastore(sqlite, mysql, pg), create an app, shut down the Fn server.
3. checkout to this branch and build the binary.
4. start the Fn server
this is what you're going to get:
```
INFO[2018-02-28T22:35:37+02:00] datastore dialed                              datastore=sqlite3 max_idle_connections=256
INFO[2018-02-28T22:35:37+02:00] goose: current datastore version: 0          
DEBU[2018-02-28T22:35:37+02:00] old migration table version was: 7           
OK    1_add_route_created_at.go
OK    2_add_call_stats.go
OK    3_add_call_error.go
OK    4_add_route_updated_at.go
OK    5_add_app_created_at.go
OK    6_add_app_updated_at.go
OK    7_add_route_cpus.go
DEBU[2018-02-28T22:35:37+02:00] Migrations left to apply: 0                  
```
It would be the same for mysql and postgres.